beacon-chain/rpc/validator/exit.go

```diff
@@ -0,0 +1,75 @@
+package validator
+
+import (
+	"context"
+
+	ptypes "github.com/gogo/protobuf/types"
+	ethpb "github.com/prysmaticlabs/ethereumapis/eth/v1alpha1"
+	"github.com/prysmaticlabs/go-ssz"
+	"github.com/prysmaticlabs/prysm/beacon-chain/core/helpers"
+	"github.com/prysmaticlabs/prysm/beacon-chain/core/statefeed"
+	"github.com/prysmaticlabs/prysm/shared/bls"
+	"github.com/prysmaticlabs/prysm/shared/mathutil"
+	"github.com/prysmaticlabs/prysm/shared/params"
+	"github.com/prysmaticlabs/prysm/shared/roughtime"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// RequestExit requests an exit for a validator.
+func (vs *Server) RequestExit(ctx context.Context, req *ethpb.VoluntaryExit) (*ptypes.Empty, error) {
+	s, err := vs.HeadFetcher.HeadState(ctx)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not get head state: %v", err)
+	}
+
+	// Confirm the validator is eligible to exit with the parameters provided
+	if req.ValidatorIndex >= uint64(len(s.Validators)) {
+		return nil, status.Errorf(codes.InvalidArgument, "Unknown validator index %d", req.ValidatorIndex)
+	}
+	validator := s.Validators[req.ValidatorIndex]
+	if !helpers.IsActiveValidator(validator, req.Epoch) {
+		return nil, status.Errorf(codes.InvalidArgument, "Validator %d not active at epoch %d", req.ValidatorIndex, req.Epoch)
+	}
+	if validator.ExitEpoch != params.BeaconConfig().FarFutureEpoch {
+		return nil, status.Errorf(codes.InvalidArgument, "Validator %d already exiting", req.ValidatorIndex)
+	}
+
+	secondsPerEpoch := params.BeaconConfig().SecondsPerSlot * params.BeaconConfig().SlotsPerEpoch
+	genesisTime := vs.GenesisTimeFetcher.GenesisTime().Unix()
+	currentEpoch := uint64(roughtime.Now().Unix()-genesisTime) / secondsPerEpoch
+	earliestRequestedExitEpoch := mathutil.Max(req.Epoch, currentEpoch)
+	earliestExitEpoch := validator.ActivationEpoch + params.BeaconConfig().PersistentCommitteePeriod
+	if earliestRequestedExitEpoch < earliestExitEpoch {
+		return nil, status.Errorf(codes.InvalidArgument, "Validator %d cannot exit before epoch %d", req.ValidatorIndex, earliestExitEpoch)
+	}
+
+	// Confirm signature is valid
+	root, err := ssz.SigningRoot(req)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, "Malformed request")
+	}
+	sig, err := bls.SignatureFromBytes(req.Signature)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "Malformed signature: %v", err)
+	}
+	validatorPubKey, err := bls.PublicKeyFromBytes(validator.PublicKey)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Invalid validator public key: %v", err)
+	}
+	domain := bls.ComputeDomain(params.BeaconConfig().DomainVoluntaryExit)
+	verified := sig.Verify(root[:], validatorPubKey, domain)
+	if !verified {
+		return nil, status.Error(codes.InvalidArgument, "Incorrect signature")
+	}
+
+	// Send the voluntary exit to the state feed.
+	vs.StateNotifier.StateFeed().Send(&statefeed.Event{
```

Comment:
What's the reasoning for sending the exit to the state feed?  I think it would just be broadcast to the network here

Reply:
We need to keep track of the exits locally so we can add them to proposed blocks (and may also involve some sort of DB storage to keep them around in case of a crash; haven't thought that piece through entirely).

Rough idea is there will be a service that receives voluntary exits, stores them in the local pool, then gossips them out, so the broadcast wouldn't be until after we've stored it.  Otherwise we'll end up just passing the same requests round and round the network.

Reply:
I see. A few notes:

1. What do you think about having a separate `blockFeed`?. `stateFeed` can just be `processed_state` or `head_state`. I like the separation 
2. We do need to broadcast the exit object here, we should either add a TODO or add the broadcast method
3. I'm refactoring the block operations object pool for `v0.9.2`(#4191) What do you think about having a similar pool for exit? We can use a less fancy cache since expirations don't matter (i think)
4. Let's change the base from `master` to `v0.9.2`, as this ties in better with all the `v0.9.2` features

Reply:
1. We always have the question of how generic we make the feed.  Single-use feeds end up proliferating and being a pain to look after, but `stateFeed` isn't a one-size-fits all either.  Perhaps have a `txFeed` that handles attestations and voluntary exits (and extending it to transfers and the like as they come in to the frame), and leave `stateFeed` for notifications of changes to the state of the chain?  That also allows us to size subscriber channels to `txFeed` appropriately, as they may have more of a backlog of events to handle due to needing to handle them sequentially.
2. I'm unsure why we'd broadcast before checking if we have it in the pool (and hence already broadcast it); without that check someone could use this as a DDOS vector
3. Very much yes.  If we can reuse the same logic (albeit separate pools) for attestations and voluntary exits (and again for transfers etc.) that would be most handy.
4. Willdo.

Reply:
1. Sounds good. I'll make an `operationFeed` today. Given `tx` is a reserved keyword for users level objections 😅 
2. This is what we do with attestations. In the current architecture there's trust assumption between beacon node and validator as one entity. Ex: you wouldn't run your beacon node to service unknown validator. For service providers, there will be middlewears to whitelist validators for RPC.  
Broadcasting right away and not withholding the object for altruism since we don't know when connected validators will be able to propose block next  
3. Awesome!
4. Thanks!

Reply:
Latest commit uses operation feed for exits.